### ISSUE TITLE
Using FP8 inference in long response vllm rollouts 

### ISSUE THREAD

**Issue by vadimkantorov** [verl-project/verl#1803]
Hi! In our experience, vllm rollouts use 70% of the grpo iteration time (when performed in bf16)

Has anyone tried using more aggressive precision reduction for rollouts (like using FP8) for gaining speed? I wonder if some okay methods exist for online FP8 usage for gaining speed in this rollout phase

Thanks!

**Comment by vermouth1992** [comment]
The real issue was that the current vllm implementation doesn't support loading bf16 weights online and dynamically quantize them to fp8. We need this feature to land in vllm first and we only need to set the precision in vllm engine without modifying any code.

**Comment by vermouth1992** [comment]
Another approach is to perform quantization in the sharding manager. But we haven't got a chance to try it

**Comment by vadimkantorov** [comment]
> Another approach is to perform quantization in the sharding manager.

Does verl do weight consolidation on one rank prior to syncing them to the vllm workers?

Probably an example is needed of scaling out the rollout phase on a separate pool of machines (and thus of efficient weight broadcast/online fp8 quantization)... Does the sharding manager handle the weight consolidation/broadcast on the verl side?

If verl uses sth like https://github.com/vllm-project/vllm/blob/ebb1ec931871ee1baca23673049865856c44ce4e/examples/offline_inference/rlhf_utils.py#L53, could it somehow reuse some fp8 quant util from vllm and construct the fp8 weights prior to calling load_weights? Although it would be more robust if vllm supported this directly...

**Comment by vermouth1992** [comment]
> > Another approach is to perform quantization in the sharding manager.
> 
> Does verl do weight consolidation on one rank prior to syncing them to the vllm workers?
> 
> Probably an example is needed of scaling out the rollout phase on a separate pool of machines (and thus of efficient weight broadcast/online fp8 quantization)... Does the sharding manager handle the weight consolidation/broadcast on the verl side?
> 
> If verl uses sth like https://github.com/vllm-project/vllm/blob/ebb1ec931871ee1baca23673049865856c44ce4e/examples/offline_inference/rlhf_utils.py#L53, could it somehow reuse some fp8 quant util from vllm and construct the fp8 weights prior to calling load_weights? Although it would be more robust if vllm supported this directly...

It's actually here. https://github.com/volcengine/verl/blob/main/verl/workers/sharding_manager/fsdp_vllm.py#L289. verl does weight allgather per tensor to avoid OOM.

**Comment by vermouth1992** [comment]
You can write a quantization function like
```python
def quantize_to_fp8(name, param):
    return xxx
```
and modify this line to call the function
```python
quantize_to_fp8(name, param.to(device, non_blocking=True).full_tensor())
```

**Comment by vadimkantorov** [comment]
I wonder if any vllm built-in quant functions exist for this... And I also wonder if at `LLM(..., quantization="fp8")` any full-model calibration is performed to tweak the quantized weights to match the full perf (e.g. layerwise) - but of course having a basic (but still working in reasoning generation) method would already have big impact in terms of GRPO workload speeds

**Comment by vermouth1992** [comment]
Agreed. The online quantization should be implemented inside the inference engine to support a class of models and keep the API clean in RL framework.

**Comment by vadimkantorov** [comment]
> keep the API clean in RL framework.

Then I also wonder: could param allgather be fused with some fp8 quantization (to reduce the allgather param memory/communication overweight)?

It also appears that the code in https://github.com/volcengine/verl/blob/main/verl/workers/sharding_manager/fsdp_vllm.py#L289 is specialized for FSDP1?

**Comment by engint-cerebras** [comment]
Imho, you can potentially do something similar to https://github.com/NVIDIA/NeMo-RL/blob/99ba9a130e72cbf87a3e20acf43bc01a47adc8ee/nemo_rl/models/generation/vllm_backend.py

**Comment by LiyuanLucasLiu** [comment]
We have an implementation for this and I'm doing some explorations on the impact of using quantization during rollout. Currently it seems the recipe needs to be re-tuned. I will release the toolkit after finish the validation. stay tuned : )

**Comment by vadimkantorov** [comment]
> We have an implementation for this and I'm doing some explorations on the impact of using quantization during rollout.

I also wonder if fp8-quantized variant can have sufficient precision for reference model log-prob evaluation? And if also vllm can be used for ref-model inference.

**Comment by vadimkantorov** [comment]
> We have an implementation for this

I wonder if this recently merged vllm PR is helpful for this work on verl side:
- https://github.com/vllm-project/vllm/pull/18745

**Comment by hxdtest** [comment]
I have hooked vllm `_process_weights_after_loading` method  to test fp8 rollout，https://github.com/volcengine/verl/compare/main...hxdtest:FP8_RL:add_fp8#diff-7f393140ded33f50195f94ff860ef8a6e261846a1b1272a84fcc7a8f1dedbda8

**Comment by vadimkantorov** [comment]
@vermouth1992 Could you advise if:

1. Does/can Verl deallocates vllm's KV-cache memory at the end of rollout stage?
2. Does/can Verl deallocates vllm's model weights at the end of rollout stage?
3. Does/can Verl use FP8 for ref-model? Can Verl use vllm for ref-log-probs computation? Should it be faster than FSDP inference?

These can be important for the co-location setup and having more predictable memory usage

**Comment by vadimkantorov** [comment]
Also, would `model = model_class(vllm_config=vllm_config, prefix=prefix)` trigger re-torch.compile/CUDA-graph capture?

**Comment by vadimkantorov** [comment]
@hxdtest I think it's very important to merge it upstream to Verl. And then maybe even add an option for FP4 quantization.

Then I also wonder if it's possible to run ref-model also in FP8

**Comment by LiyuanLucasLiu** [comment]
we just released FlashRL for conducting quantized rollouts in RL training and identified one key technique, i.e., TIS, for stabilizing training. you can read more at: https://fengyao.notion.site/flash-rl : )

**Comment by vadimkantorov** [comment]
https://github.com/yaof20/Flash-RL seems still private. Please post your changes to vllm :) At least to this issue:
- https://github.com/vllm-project/vllm/issues/19020

Also curious, have you experimented with running the ref-model in fp8 to get the ref log-probs?

**Comment by LiyuanLucasLiu** [comment]
> https://github.com/yaof20/Flash-RL seems still private. Please post your changes to vllm :) At least to this issue:
> 
> * [[Feature]: online bf16->fp8 (and maybe even fp4?) quantization in `load_weights(...)` (so not only from disk checkpoint, but importantly from VRAM or RAM) for weights reloading in GRPO rollout workloads, and speed gains for generating long reasoning rollouts and peak VRAM reduction vllm-project/vllm#19020](https://github.com/vllm-project/vllm/issues/19020)
> 
> Also curious, have you experimented with running the ref-model in fp8 to get the ref log-probs?

Please check again, the github should be public now! As to your question, no, we haven't tried that yet.

**Comment by vadimkantorov** [comment]
https://github.com/yaof20/Flash-RL/blob/main/flash_rl/vllm_patch.py has quite a lot of patches - could you please add more comments why every function in `vllm_patch.py` is needed and more comments what exactly they fix? thank you :)

**Comment by LiyuanLucasLiu** [comment]
Right, engineering-wise, it is not easy to make it work... we wrote a guide on the high level of how FlashEL works as in: https://github.com/yaof20/Flash-RL/tree/main/tutorial

In the long run, we plan to pr this into vLLM / sglang, but don't have a concrete timeline for that yet.

**Comment by vadimkantorov** [comment]
Is the culprit that vllm does something not very optimized during vanilla load_weights? E.g. it does not use copy_?

Also, what is the contents of the "state" of Record1 and Record2? 

Thanks!

**Comment by LiyuanLucasLiu** [comment]
It cannot use `copy_`, since usually the target tensor has different shape & dtype from the source tensor, for quantized models. For example, the `marlin` kernel, if used, saves weights in a striped manner (you can check the `https://github.com/IST-DASLab/marlin` for more details). Also, particularly for `fp8`, its storage is actually `int8` (so after loading the model weight to fp8, you may find the parameter tensor to be in `int8` instead), and needs to be reinterpreted by the kernel. So its a really complicated process.

**Comment by vadimkantorov** [comment]
@LiyuanLucasLiu by chance, do you have anywhere repro pseudo/code for Figure 1 in https://fengyao.notion.site/off-policy-rl ? Are you measuring the discrepancy during learning? Or do you have by chance a standalone validation script demonstrating the discrepancy (e.g. with a fixed stock qwen) - maybe in Flash-RL repo?

A friend of mine has figured out a PyTorch hack to make vllm logprobs match fsdp logprobs and he would like to validate his results with your script.

**Comment by LiyuanLucasLiu** [comment]
> [@LiyuanLucasLiu](https://github.com/LiyuanLucasLiu) by chance, do you have anywhere repro pseudo/code for Figure 1 in https://fengyao.notion.site/off-policy-rl ? Are you measuring the discrepancy during learning? Or do you have by chance a standalone validation script demonstrating the discrepancy (e.g. with a fixed stock qwen) - maybe in Flash-RL repo?
> 
> A friend of mine has figured out a PyTorch hack to make vllm logprobs match fsdp logprobs and he would like to validate his results with your script.

I just checked in the utils we used for logging as in: https://github.com/yaof20/verl/commit/7c5954ba938f255669730eeb21e7b6584426394f The script for the `bf16` run can be found at: https://github.com/yaof20/verl/blob/flash-rl/recipe/flash_rl/dapo_qwen32b_bf16.sh

Note that we used gen_tp=2, sp_size=8, use_dynamic_bsz=True, the V1 version of the vLLM engine, and observe that the distribution gap is mostly obvious for large models and long responses. Also, we observed the `vllm_kl` to be actually small, while the `rollout_probs_diff_max` and `rollout_probs_diff_min` to be large. Additionally, MoE models like Qwen3-MoE would have a larger distribution gap than dense models. 

Looking forward to hearing good news from you guys : )

**Comment by jerryzh168** [comment]
added a PR with torchao here: https://github.com/volcengine/verl/pull/3084

**Comment by jagmarques** [comment]
The off-policy distribution gap documented in the FlashRL work (and the logprob discrepancy between vllm and FSDP) points to a deeper issue: weight quantization for rollout speed is orthogonal to KV cache memory, and the two interact during long-response generation. For 32K+ token reasoning traces, the KV cache can easily exceed the weight memory footprint, meaning FP8 weight quantization (which targets weight memory) helps throughput but the KV buffer itself stays FP16.

A dimension worth exploring alongside FP8 weight quantization: **KV cache compression during rollout**. If you compress cached KV entries to 3–4 bits training-free (no calibration), you get 4–5x more effective context per GPU without the allgather communication overhead changing. The allgather in  is per-weight-tensor; a smaller KV footprint doesn't reduce that but it reduces the per-step peak memory so more sequences fit in a batch, increasing rollout throughput.

The FlashRL approach of hooking  is clean. A KV compression hook would sit at the vllm attention backend level — compress after each prefill step, decompress at decode attention. The key advantage for RL rollouts specifically: you don't need the full-precision KV for subsequent tokens, just for the log-prob computation at the end, so lossy compression with ~0.03% perplexity impact is acceptable.

Has anyone benchmarked the memory breakdown during a full GRPO iteration — weights vs. KV cache vs. activations — at 32K response length? That would clarify whether KV or weight memory is the binding constraint at scale.
